onnxruntime/core/providers/cpu/tensor/upsample.h

```diff
@@ -31,33 +41,162 @@ class UpsampleBase {
+    std::string coordinate_transform_mode = start > 10
+                                                ? info.GetAttrOrDefault<std::string>("coordinate_transformation_mode", "half_pixel")
+                                                : "asymmetric";
+    get_original_coordinate_ = GetOriginalCoordinateFromResizedCoordinate(coordinate_transform_mode);
+    use_extrapolation_ = need_roi_input_ = coordinate_transform_mode == "tf_crop_and_resize" ? true : false;
+
+    std::string nearest_mode = info.GetAttrOrDefault<std::string>("nearest_mode", "round_prefer_floor");
+    get_nearest_pixel_ = GetNearestPixelFromOriginal(nearest_mode, start);
+
+    cubic_coeff_a_ = info.GetAttrOrDefault<float>("cubic_coeff_a", -0.75f);
+    exclude_outside_ = info.GetAttrOrDefault<int64_t>("exclude_outside", 0) == 0 ? false : true;
+
+    // after version 11 update, this optimization is no longer applicable for all the available modes...
+    // TODO : needs more testing to enable this for version 11
+    use_nearest2x_optimization = start > 10 ? false : true;
+
+    if (start > 10) {
+      roi_input_idx_ = 1;
+      scales_input_idx_ = 2;
+      sizes_input_idx_ = 3;
+    } else if (start <= 10 && input_count > 1) {
+      scales_input_idx_ = 1;
     }
 
-    if (input_count > 1) {
+    if (scales_input_idx_ > 0) {
       const Tensor* scale;
-      bool get_scale = info.TryGetConstantInput(1, &scale);
+      bool get_scale = info.TryGetConstantInput(scales_input_idx_, &scale);
 
       if (get_scale) {
         ParseScalesData(scale, scales_);
         scales_cached_ = true;
       }
     }
+
+    // roi is only needed when coordinate transformation mode is tf_crop_and_resize
+    // for all other modes no need to read roi input
+    if (roi_input_idx_ > 0 && need_roi_input_) {
+      const Tensor* roi;
+      bool get_roi = info.TryGetConstantInput(roi_input_idx_, &roi);
+
+      if (get_roi) {
+        ParseRoiData(roi, roi_);
+        roi_cached_ = true;
+      }
+    }
   }
 
+  GetOriginalCoordinateFunc get_original_coordinate_;
+  GetNearestPixelFunc get_nearest_pixel_;
+  float cubic_coeff_a_;
+  bool exclude_outside_;
+  float extrapolation_value_;
   UpsampleMode mode_;
+  bool use_nearest2x_optimization = false;
+
   std::vector<float> scales_;
+  std::vector<float> roi_;
   bool scales_cached_;
+  bool roi_cached_;
+  bool need_roi_input_;
+  bool use_extrapolation_;
   bool is_resize = false;
 
+  int roi_input_idx_ = -1;
+  int scales_input_idx_ = -1;
+  int sizes_input_idx_ = -1;
+
   UpsampleMode StringToUpsampleMode(const std::string& mode) {
     if (strcmp(mode.c_str(), UpsampleModeNN) == 0) {
       return UpsampleMode::NN;
     }
     if (strcmp(mode.c_str(), UpsampleModeLinear) == 0) {
       return UpsampleMode::LINEAR;
     }
-      ORT_THROW("mode attribute is " + mode + ". It can only be " +
-                UpsampleModeNN + "(default) or " + UpsampleModeLinear + ".");
+
+    if (strcmp(mode.c_str(), UpsampleModeCubic) == 0) {
```

Comment:
Do we need strcasecmp?

Reply:
Nope... it was part of old code... changed it now